### PR TITLE
[MM-53810] Add highlight support for highlight without notifications to markdown in mobile

### PR DESCRIPTION
**Pull request by M-ZubairAhmed**
#### Ticket Link
https://mattermost.atlassian.net/browse/MM-53810

#### Checklist

- [ ] Added or updated unit tests (required for all new features)
- [ ] Has UI changes
- [ ] Includes text changes and localization file updates
- [ ] Have tested against the 5 core themes to ensure consistency between them.

#### Device Information
This PR was tested on: 

#### Release Note

```release-note
Added a new feature for highlighting without notification without the ability to configure.
```

### CODE REVIEW

**Review by harshilsharma63 — APPROVED** [review]
LGTM 👍

File: app/queries/servers/system.ts [hunk 1]
```diff
@@ -543,3 +543,41 @@ export const observeLastServerVersionCheck = (database: Database) => {
...
+    );
+
+    const isEnterpriseReady = observeConfigBooleanValue(database, 'BuildEnterpriseReady', false);
+
+    return combineLatest([isCloudStarterFree, isStarterSKULicense, isEnterpriseReady, IsLicensed]).pipe(
+        switchMap(([isCSF, isSKL, isEnt, isLic]) => {
+            const isSelfHostedStarter = isEnt && isLic === false;
+            const isEnterpriseOrCloudOrSKUStarterFree = isCSF || isSKL || isSelfHostedStarter;
+
+            return of$(!isEnterpriseOrCloudOrSKUStarterFree && isEnt);
```
Comment (larkox) [inline app/queries/servers/system.ts:580]: It is a bit hard to follow all the conditions, but I cannot think of a better way to make them clearer.

Maybe some comments about what exactly are you checking? (i.e. what are the kind of licenses that have it and what not)

**Review by larkox — CHANGES_REQUESTED** [review]
In general looks good. Just marking it as `request changes`for the number of comments, but I don't think there is anything major in them.

File: app/components/post_list/post/body/message/message.tsx [hunk 2]
```diff
@@ -96,7 +93,8 @@ const Message = ({currentUser, highlight, isEdited, isPendingOrFailed, isReplyPo
                             postId={post.id}
                             textStyles={textStyles}
                             value={post.message}
-                            mentionKeys={mentionKeys}
+                            mentionKeys={currentUser?.mentionKeys ?? []}
+                            highlightKeys={isHighlightWithoutNotificationLicensed ? (currentUser?.highlightKeys ?? []) : []}
```
[the remark this answers is not in the document]
Comment (M-ZubairAhmed) [reply app/components/post_list/post/body/message/message.tsx:97]: Done ✅

File: app/components/markdown/transform.test.ts [hunk 3]
```diff
@@ -2801,6 +2803,316 @@ describe('Components.Markdown.transform', () => {
             });
         }
     });
+
+    describe('highlightWithoutNotification', () => {
```
[the remark this answers is not in the document]
Comment (M-ZubairAhmed) [reply app/components/markdown/transform.test.ts:2807]: Done ✅

File: app/queries/servers/system.ts [hunk 4]
```diff
@@ -543,3 +543,41 @@ export const observeLastServerVersionCheck = (database: Database) => {
...
+            return of$(isLicensed && isSelfHostedStarterProduct);
+        }),
+        distinctUntilChanged(),
+    );
+
+    const isEnterpriseReady = observeConfigBooleanValue(database, 'BuildEnterpriseReady', false);
+
+    return combineLatest([isCloudStarterFree, isStarterSKULicense, isEnterpriseReady, IsLicensed]).pipe(
+        switchMap(([isCSF, isSKL, isEnt, isLic]) => {
+            const isSelfHostedStarter = isEnt && isLic === false;
```
[the remark this answers is not in the document]
Comment (M-ZubairAhmed) [reply app/queries/servers/system.ts:577]: Done ✅

File: app/queries/servers/system.ts [hunk 5]
```diff
@@ -543,3 +543,41 @@ export const observeLastServerVersionCheck = (database: Database) => {
         switchMap((model) => of$(parseInt(model.value, 10))),
     );
 };
+
+export const observeIfHighlightWithoutNotificationHasLicense = (database: Database) => {
+    const IsLicensed = observeLicense(database).pipe(
```
[the remark this answers is not in the document]
Comment (M-ZubairAhmed) [reply app/queries/servers/system.ts:548]: Done ✅

[the remark this answers is not in the document]
Comment (M-ZubairAhmed) [reply app/queries/servers/system.ts:548]: Done ✅

File: app/components/markdown/transform.ts [hunk 6]
```diff
@@ -174,6 +174,55 @@ export function mentionKeysToPatterns(mentionKeys: UserMentionKey[]) {
     });
 }
 
+export function highlightWithoutNotification(ast: Node, highlightKeys: HighlightWithoutNotificationKey[]) {
+    const walker = ast.walker();
+
+    const patterns = highlightKeysToPatterns(highlightKeys);
```
[the remark this answers is not in the document]
Comment (M-ZubairAhmed) [reply app/components/markdown/transform.ts:180]: as per the markdown logic this is executed on each text node of a post

File: package-lock.json [hunk 7]
```diff
@@ -30014,9 +30015,9 @@
       "integrity": "sha1-3dgA2gxmEnOTzKWVDqloo6rxJTs="
     },
     "commonmark": {
-      "version": "npm:@mattermost/commonmark@0.30.1-1",
-      "resolved": "https://registry.npmjs.org/@mattermost/commonmark/-/commonmark-0.30.1-1.tgz",
-      "integrity": "sha512-zOiOtPA5LjB1WCUVbaMt5KNuNdq/0NLpBOcQlGrKe+ftuJVhiis7pWEqS4agIBlpjhm7ssywwl7p1eurOjcm5A==",
+      "version": "git+ssh://git@github.com/mattermost/commonmark.js.git#9ebe192ca8d2b3db4955931990d6639d21fc79de",
```
[the remark this answers is not in the document]
Comment (M-ZubairAhmed) [reply package-lock.json:30018]: i ran it again after reverting this change. Its ok now

**Review by larkox — APPROVED** [review]
Great work! Looks good to me

**Review by yasserfaraazkhan — APPROVED** [review]
Tested this on  iOS 17.0.3, Android 11.0
- able to see words and phrases highlighted.
- behaviour is same on web and mobile.
- able to edit on web and see the changes on mobile
